[zengzihui] iP

zengzihui opened the pull request:
# DukeCicada

> Hello World, here is Cicada

DukeCicada frees your mind of having to remember things you need to do. It's 
- text-based
- easy to learn
- ~~fast~~ *SUPER* FAST to use


All you need to do is,
1. download it from [here](https://github.com/zengzihui/ip/releases/tag/A-Jar)
2. double-click it
3. add your tasks
4. let it manage your tasks for you 😄 

And it is **FREE**!

Feature:
- [x] Add Task
- [x] Delete Task
- [x] Find Task
- [ ] Reminders (coming soon)
- [ ] Managing deadlines (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Duke {
    public static void main(String[] args) {
        System.out.println("Hello World");
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

zengzihui committed:
Finish Level 0. Rename, Greet, Exit

zengzihui committed:
Finish Level 1. Echo

zengzihui committed:
Finish Level 2. Add, List

zengzihui committed:
Finish Level 3. Mark as Done

zengzihui committed:
Finish Level 4. ToDos, Events, Deadlines

zengzihui committed:
Finished A-TextUiTesting

zengzihui committed:
Finish Level 5. Handle Errors

zengzihui committed:
Finished Level 6. Delete and fixed Level-5 errors (included handling errors for level 5).

zengzihui committed:
Update .gitignore

ignore the class files

zengzihui committed:
Finish Level 7. Save
Save feature is added as the requirement. Besides that,
The code is refractored to Switch statement instead of
original if-else statements. This is because when adding
new features inside, the original if-else statements become
kind of messy. Some code is rearranged into methods to make
it more modularised.

zengzihui committed:
Merge branch 'branch-Level-7'

lihaoquan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,294 @@
+        System.out.println(horizontalLine);
+    }
+
+    public static void save(Path path, List<Task> tasks) {
```
Perhaps the function name can be more descriptive.

lihaoquan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,294 @@
+            System.err.println("There are some error in saving. Try again");
+        }
+    }
+    public static void greeting(String chatbotName, String horizontalLine) {
```
Would `printGreeting` be a better alternative for this function's name?

lihaoquan commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,8 @@
+todo borrow book
+deadline return book by this monday
+event project meeting from 5pm to 7pm
+list
+mark 3
+list
+unmark 3
+list
```
Nice thorough testing of the text UI.

lihaoquan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,294 @@
+        if (number.isBlank()) {
+            throw new DukeException("Please state which task you want to delete.");
+        }
+        int index = Integer.parseInt(number) - 1;
```
A bit of nitpicking but would `deleteIndex` be a better variable name in case you are working with other index variables in the same function in the future?

lihaoquan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,294 @@
+        // String home = System.getProperty("user.home");
+        // Path path = Paths.get(home, "OneDrive","Documents", "NUS",
+        //         "Y2S2", "CS2103", "Individual Project", "duke.txt");
+        String relativePath = "data/duke.txt";
```
This can perhaps be a constant with the name "RELATIVE_PATH" since it will remain unchanged throughout.

zengzihui replied:
thankss, updated

lihaoquan reviewed (commented):
Some minor suggestions for variable names.
Generally looks good.

zengzihui committed:
Finish Level 8. Dates and Times
Replace the String variable of date time to LocalDateTime var and
add exception handling for that. Additionally, change some variable
and function name as suggested by the pull request peer code review.

zengzihui committed:
Finish A-MoreOOP
Change the code structure by adding some packages and add respective
new classes to let it be more OOP. The code is structured by
its purpose. such as commands, constants, exceptions, parsers,
storage, tasks, ui.

zengzihui committed:
Finish A-Packages
organise the code into packages to make it more neat.

zengzihui committed:
Finish A-Packages
Commit those files that left out just now

zengzihui committed:
Merge remote-tracking branch 'origin/add-gradle-support'

zengzihui committed:
Finish A-JUnit
Create 2 test file for 2 classes, Parser and TaskList
Test the empty and invalid input for mark task
and create deadline task.

zengzihui committed:
Finish Level 9. Find
Implement the search() method in Task class and find() method
in TaskList class to allow user to find a task by searching for
a keyword

zengzihui committed:
Finish A-JavaDoc
Add the Java Doc to every class and methods.

zengzihui committed:
Finish A-CodingStandard
Tweak the code to comply with Java Coding Standard.

zengzihui committed:
Merge branch-A-CodingStandard
Merge them.

zengzihui committed:
Merge branch-A-JavaDoc to master
Merge them.

Howlong11 commented on src/main/java/duke/commands/FindCommand.java:
Javadoc comments for this class are missing

Howlong11 commented on src/main/java/duke/constants/Constant.java:
I like that you created a class for constants. 

However, I think it could be better to name the DateTimeFormatters as INPUT_DATE_TIME_FORMATTER and OUTPUT_DATE_TIME_FORMATTER just so that it is easier to understand which is which.

Howlong11 commented on src/main/java/duke/commands/HelpCommand.java:
I like that you created help command for possible commands!

Howlong11 commented on src/main/java/duke/parsers/Parser.java:
Perhaps you could use line breaks for your error messages...?

Howlong11 commented on src/main/java/duke/ui/Ui.java:
Your new logo looks cool 👍

zengzihui replied:
Thanks

zengzihui committed:
Finish A-CheckStyle
Add the checktyle config files and add the plugin

zengzihui committed:
Finish Level 10. GUI

Add a JavaFX GUI to my Cicada chatbot. Change the display methods
to return String.

zengzihui committed:
Update build.gradle file to generate JAR file

Change the mainClass to duke.Launcher for generating JAR file for this
new JavaFX GUI. Comment out the JUnit tests as now it is no longer
Text-UI based, and the methods have been modified for the JavaFX GUI
already.

zengzihui committed:
Add assertions to Duke and MainWindow class

Use assert feature to document some important assumptions made in
the Duke class and MainWindow class. I was stuck at null Ui object
for previous task, hence I assert Ui != null in the Duke class.

zengzihui committed:
Shorten prepareCreateEvent() method in Parser class

Create isTaskArgumentBlank() and convertToLocalDateTime() method in Parser class
to shorten line of codes and following the code quality guideline of no more than
30 lines of code for a method.

zengzihui committed:
Shorten parse() method in Parser class

Convert the handling of each command word into individual methods
so that the line of codes for this method will be within 30 lines.

zengzihui committed:
Shorten load() method in Storage class

Convert part of code in load() to a new method called addTaskToList()
to shorten the line of codes in load() method as is exceeds 30 lines.
Change the code structure of load() to minimise deep nesting.

zengzihui committed:
Add Utils method to put common methods

Add Utils class to put common methods so that these methods
can be used among different classes.

zengzihui committed:
Add Javadoc for all classes

Add JavaDoc for all methods and classes.

zengzihui committed:
Merge pull request #3 from zengzihui/branch-A-CodeQuality

Improve code quality

zengzihui committed:
Merge branch 'master' into branch-A-Assertions

zengzihui committed:
Merge pull request #2 from zengzihui/branch-A-Assertions

Add assertions to Duke and MainWindow class

zengzihui committed:
Finish B-ViewSchedules feature

Implement View Schedules to allow user to view tasks in the form of a schedule,
users are able to view the schedule for a specified date.

zengzihui committed:
Change manually save to automatically save

Update the save feature from manually save to automatically save
as I did not realised the requirement is auto save when I do this task

zengzihui committed:
Modify the JUnit tests to test the new methods

As the methods for JUnit testing have been updated after adding the GUI,
hence the JUnit tests are updated to test the updated methods.

zengzihui committed:
Update user guide

Update user guide to teach user how to use Cicada.
